tests/run-make/mixing-formats/rmake.rs

```diff
@@ -0,0 +1,87 @@
+// Testing various mixings of rlibs and dylibs. Makes sure that it's possible to
+// link an rlib to a dylib. The dependency tree among the file looks like:
+//
+//                 foo
+//               /     \
+//             bar1   bar2
+//             /    \ /
+//          baz    baz2
+//
+// This is generally testing the permutations of the foo/bar1/bar2 layer against
+// the baz/baz2 layer
+
+//@ ignore-cross-compile
+
+use run_make_support::{rustc, tmp_dir};
+use std::fs;
+
+fn main() {
+    // Building just baz
+    rustc().crate_type("rlib").input("foo.rs").run();
+    rustc().crate_type("dylib").input("bar1.rs").arg("-Cprefer-dynamic").run();
+    rustc().crate_type("dylib,rlib").input("baz.rs").arg("-Cprefer-dynamic").run();
+    rustc().crate_type("bin").input("baz.rs").run();
+    fs::remove_dir_all(tmp_dir()).unwrap();
+    fs::create_dir(tmp_dir()).unwrap();
```

Comment:
Suggestion: it might be cleaner to create a helper to abstract the "clean `tmp_dir()` functionality, i.e.

```rs
// helper
fn test_with_teardown(f: Fn()) {
    f();
    fs::remove_dir_all(tmp_dir()).unwrap();
    fs::create_dir(tmp_dir()).unwrap();
}

// callsite
fn main() {
    test_with_teardown(|| {
        // Building just baz
        rustc().crate_type("rlib").input("foo.rs").run();
        rustc().crate_type("dylib").input("bar1.rs").arg("-Cprefer-dynamic").run();
        rustc().crate_type("dylib,rlib").input("baz.rs").arg("-Cprefer-dynamic").run();
        rustc().crate_type("bin").input("baz.rs").run();
    });
}
```

Reply:
You're right, it's a lot cleaner! I wanted to abstract over this test a lot initially, making a Flags struct with fields `crate_type`, `prefers_dynamic` and `should_fail`, but I think writing everything out is clearer... with the slight improvement of your helper teardown function. Added.